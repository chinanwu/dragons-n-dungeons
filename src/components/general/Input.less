.Input {
  background-color: inherit;
  border: none;
  width: 97%;
}
