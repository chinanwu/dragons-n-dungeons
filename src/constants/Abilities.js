export const ABILITY_NAMES = {
  strength: 'Strength',
  str: 'Strength',
  constitution: 'Constitution',
  con: 'Constitution',
  dexterity: 'Dexterity',
  dex: 'Dexterity',
  wisdom: 'Wisdom',
  wis: 'Wisdom',
  intelligence: 'Intelligence',
  int: 'Intelligence',
  charisma: 'Charisma',
  cha: 'Charisma',
};

export const STRENGTH_SKILLS = ['Saving Throw', 'Athletics'];
export const CONSTITUTION_SKILLS = ['Saving Throw'];
export const DEXTERITY_SKILLS = [
  'Saving Throw',
  'Acrobatics',
  'Stealth',
  'Sleight of Hand',
];
export const WISDOM_SKILLS = [
  'Saving Throw',
  'Animal Handling',
  'Insight',
  'Medicine',
  'Perception',
  'Survival',
];
export const INTELLIGENCE_SKILLS = [
  'Saving Throw',
  'Arcana',
  'History',
  'Investigation',
  'Nature',
  'Religion',
];
export const CHARISMA_SKILLS = [
  'Saving Throw',
  'Deception',
  'Intimidation',
  'Performance',
  'Persuasion',
];

export const ABILITY_SKILLS = {
  strength: STRENGTH_SKILLS,
  str: STRENGTH_SKILLS,
  constitution: CONSTITUTION_SKILLS,
  con: CONSTITUTION_SKILLS,
  dexterity: DEXTERITY_SKILLS,
  dex: DEXTERITY_SKILLS,
  wisdom: WISDOM_SKILLS,
  wis: WISDOM_SKILLS,
  intelligence: INTELLIGENCE_SKILLS,
  int: INTELLIGENCE_SKILLS,
  charisma: CHARISMA_SKILLS,
  cha: CHARISMA_SKILLS,
};
