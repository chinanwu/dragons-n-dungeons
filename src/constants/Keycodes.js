// Arrow keys
export const leftArrow = 37;
export const upArrow = 38;
export const rightArrow = 39;
export const downArrow = 40;
