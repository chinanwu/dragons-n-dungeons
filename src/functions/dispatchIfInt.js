import isNumber from './isNumber';

export default (value, actionFn) => dispatch => {
  const test = isNumber(value);
  test !== null ? dispatch(actionFn(test)) : null;
};
