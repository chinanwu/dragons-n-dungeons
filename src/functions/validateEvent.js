export default (event, additional) => event && event.target && additional;
