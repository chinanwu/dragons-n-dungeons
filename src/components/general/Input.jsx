import PropTypes from 'prop-types';
import React, { useCallback } from 'react';

import { downArrow, upArrow } from '../../constants/Keycodes';
import dispatchIfInt from '../../functions/dispatchIfInt';
import validateEvent from '../../functions/validateEvent';

import './Input.less';

export const Input = ({
  id,
  className,
  type,
  defaultInput,
  ariaChecked,
  ariaLabel,
  ariaLabelledBy,
  onChange,
  onKeyDown,
  ...props
}) => {
  const handleKeyDown = useCallback(
    event => {
      if (validateEvent(event)) {
        if (event.target.value && event.keyCode) {
          switch (event.keyCode) {
            case upArrow:
              event.preventDefault();
              dispatchIfInt(event.target.value, value => value + 1)(onKeyDown);
              break;
            case downArrow:
              event.preventDefault();
              dispatchIfInt(event.target.value, value => value - 1)(onKeyDown);
              break;
            default:
              onChange(event);
              break;
          }
        }
      }
    },
    [onChange, onKeyDown]
  );

  // There is a bug with react where when an input is type='number', it allows
  // for zero-padded numbers - e.g. 01. I want to always have those numbers be
  // truncated to just be e.g. 1. Only way to do that is to unfortunately use
  // input type='text'
  return (
    <input
      id={id}
      className={'Input ' + className}
      type={type === 'text' || type === 'number' ? 'text' : type}
      aria-checked={ariaChecked}
      aria-label={ariaLabel}
      aria-labelledby={ariaLabelledBy}
      onChange={onChange}
      onKeyDown={type === 'number' ? handleKeyDown : null}
      {...props}
    />
  );
};

Input.propTypes = {
  id: PropTypes.string.isRequired,
  className: PropTypes.string,
  type: PropTypes.string,
  defaultInput: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  ariaChecked: PropTypes.bool,
  ariaLabel: PropTypes.string,
  ariaLabelledBy: PropTypes.string,
  onChange: PropTypes.func,
  onKeyDown: PropTypes.func,
};

export default Input;
