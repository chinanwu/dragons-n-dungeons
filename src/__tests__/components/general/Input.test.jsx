import { shallow } from 'enzyme';
import React from 'react';

import { Input } from '../../../components/general/Input.jsx';
import { downArrow, leftArrow, upArrow } from '../../../constants/Keycodes';
import dispatchIfInt from '../../../functions/dispatchIfInt';
import validateEvent from '../../../functions/validateEvent';

jest.unmock('../../../components/general/Input.jsx');
jest.unmock('../../../constants/Keycodes');

describe('Input component', () => {
  describe('rendering', () => {
    it('renders Input', () => {
      const wrapper = shallow(<Input id="test" />);
      expect(wrapper.getElement()).toMatchSnapshot();
    });

    it('renders Input as text when number', () => {
      const wrapper = shallow(<Input id="test" type="number" />);
      expect(wrapper.getElement()).toMatchSnapshot();
      expect(wrapper.find('#test').prop('type')).toEqual('text');
    });
  });

  describe('behaviour', () => {
    it('increases value by 1 when up is pressed ', () => {
      validateEvent.mockReturnValue(true);
      dispatchIfInt.mockImplementation(() => () => {});
      const preventDefault = jest.fn();
      const event = {
        target: {
          value: '1',
        },
        keyCode: upArrow,
        preventDefault,
      };

      const wrapper = shallow(
        <Input id="test" type="number" onKeyDown={jest.fn()} />
      );
      expect(wrapper.getElement()).toMatchSnapshot();
      wrapper.find('#test').simulate('keydown', event);
      expect(preventDefault).toHaveBeenCalled();
    });

    it('decreases value by 1 when down is pressed ', () => {
      validateEvent.mockReturnValue(true);
      dispatchIfInt.mockImplementation(() => () => {});
      const preventDefault = jest.fn();
      const event = {
        target: {
          value: '1',
        },
        keyCode: downArrow,
        preventDefault,
      };

      const wrapper = shallow(
        <Input id="test" type="number" onKeyDown={jest.fn()} />
      );
      expect(wrapper.getElement()).toMatchSnapshot();
      wrapper.find('#test').simulate('keydown', event);
      expect(preventDefault).toHaveBeenCalled();
    });

    it('calls onChange when any other key is pressed', () => {
      validateEvent.mockReturnValue(true);
      dispatchIfInt.mockImplementation(() => () => {});
      const preventDefault = jest.fn();
      const onChange = jest.fn();
      const event = {
        target: {
          value: '1',
        },
        keyCode: leftArrow,
        preventDefault,
      };

      const wrapper = shallow(
        <Input
          id="test"
          type="number"
          onKeyDown={jest.fn()}
          onChange={onChange}
        />
      );
      expect(wrapper.getElement()).toMatchSnapshot();
      wrapper.find('#test').simulate('keydown', event);
      expect(preventDefault).not.toHaveBeenCalled();
      expect(onChange.mock.calls).toEqual([[event]]);
    });

    it('does nothing with invalid event', () => {
      validateEvent.mockReturnValue(false);
      const wrapper = shallow(<Input id="test" type="number" />);
      expect(wrapper.getElement()).toMatchSnapshot();
      wrapper.find('#test').simulate('keydown', null);
    });

    it('does nothing with event that has no value', () => {
      validateEvent.mockReturnValue(true);
      const event = {
        target: {},
      };
      const wrapper = shallow(<Input id="test" type="number" />);
      expect(wrapper.getElement()).toMatchSnapshot();
      wrapper.find('#test').simulate('keydown', event);
    });
  });
});
