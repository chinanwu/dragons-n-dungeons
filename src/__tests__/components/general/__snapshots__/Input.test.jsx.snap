// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component behaviour calls onChange when any other key is pressed 1`] = `
<input
  className="Input undefined"
  id="test"
  onChange={[MockFunction]}
  onKeyDown={[Function]}
  type="text"
/>
`;

exports[`Input component behaviour decreases value by 1 when down is pressed  1`] = `
<input
  className="Input undefined"
  id="test"
  onKeyDown={[Function]}
  type="text"
/>
`;

exports[`Input component behaviour does nothing with event that has no target 1`] = `
<input
  className="Input undefined"
  id="test"
  onKeyDown={[Function]}
  type="text"
/>
`;

exports[`Input component behaviour does nothing with event that has no value 1`] = `
<input
  className="Input undefined"
  id="test"
  onKeyDown={[Function]}
  type="text"
/>
`;

exports[`Input component behaviour does nothing with invalid event 1`] = `
<input
  className="Input undefined"
  id="test"
  onKeyDown={[Function]}
  type="text"
/>
`;

exports[`Input component behaviour increases value by 1 when up is pressed  1`] = `
<input
  className="Input undefined"
  id="test"
  onKeyDown={[Function]}
  type="text"
/>
`;

exports[`Input component rendering renders Input 1`] = `
<input
  className="Input undefined"
  id="test"
  onKeyDown={null}
/>
`;

exports[`Input component rendering renders Input as text when number 1`] = `
<input
  className="Input undefined"
  id="test"
  onKeyDown={[Function]}
  type="text"
/>
`;
